lago/plugins/vm.py

```diff
@@ -246,36 +250,97 @@ def extract_paths(self, paths, ignore_nopath):
+            "tar "
+            "--dereference "
+            "--no-same-permissions "
+            "--no-same-owner "
+            "-c {} | gzip -f -{}"
+        )
+        remote_paths_arg = " ".join(remote_paths)
+        return cmd.format(remote_paths_arg, compression_level)
+
+    def _pipe_ssh_cmd_output_to_file(self, cmd, out_file):
+        with self.vm._ssh() as client:
+            _, stdout, _ = client.exec_command(cmd)
+            out_file.write(stdout.read())
+            out_file.flush()
+            os.fsync(out_file.fileno())
+
+    @contextlib.contextmanager
+    def _tar_gz_archive_from(self, remote_paths, compression_level=5):
+        with tempfile.NamedTemporaryFile() as archive_file:
+            tar_gz_cmd = self._prepare_tar_gz_command(
+                remote_paths, compression_level
+            )
+            self._pipe_ssh_cmd_output_to_file(tar_gz_cmd, archive_file)
             LOGGER.debug(
-                'Extracting scp://%s:%s to %s',
-                self.vm.name(),
-                host_path,
-                guest_path,
+                "Created %s archive with collected paths", archive_file.name
             )
-            try:
-                self.vm.copy_from(
-                    local_path=guest_path,
-                    remote_path=host_path,
-                    propagate_fail=not ignore_nopath
+            yield archive_file
+
+    @contextlib.contextmanager
+    def _remote_paths_extracted_to_temp_dir(self, remote_paths):
+        with self._tar_gz_archive_from(remote_paths) as archive_file:
+            with self._make_temporary_directory() as tmpdir_path:
+                LOGGER.debug(
+                    "Extracting archive %s to temporary directory: %s",
+                    archive_file.name, tmpdir_path
                 )
-            except ExtractPathNoPathError as err:
-                if ignore_nopath:
-                    LOGGER.debug(
-                        '%s: ignoring since ignore_nopath was set to True',
-                        err.args[0]
+                args = ["tar", "-xf", archive_file.name, "-C", tmpdir_path]
+                try:
+                    subprocess.check_output(args, stderr=subprocess.STDOUT)
+                except subprocess.CalledProcessError as e:
+                    LOGGER.error("'tar' command failed: %s", e.output)
+                yield tmpdir_path
+
+    @check_running
+    def _extract_paths_tar_gz(self, paths, ignore_nopath):
+        remote_paths = tuple(p[0] for p in paths)
+
+        with self._remote_paths_extracted_to_temp_dir(
+            remote_paths
+        ) as tmpdir_path:
+            for remote_path, desired_local_path in paths:
+                LOGGER.debug(
+                    'Moving %s to %s',
+                    remote_path,
+                    desired_local_path,
+                )
+                try:
+                    current_local_path = os.path.join(
+                        tmpdir_path, remote_path[1:]
```

Comment:
Why do you need to remove the first char from `remote_path`?

Reply:
To make 'os.path.join' work - 'remote_path' is absolute. This indeed seems cryptic so I've added a comment explaining what's happening.